graphite/graphite.go

```diff
@@ -0,0 +1,188 @@
+package graphite
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdb/influxdb"
+)
+
+var (
+	// ErrBindAddressRequired is returned when starting the Server
+	// without a TCP or UDP listening address.
+	ErrBindAddressRequired = errors.New("bind address required")
+
+	// ErrServerClosed return when closing an already closed graphite server.
+	ErrServerClosed = errors.New("server already closed")
+
+	// ErrDatabaseNotSpecified retuned when no database was specified in the config file
+	ErrDatabaseNotSpecified = errors.New("database was not specified in config")
+
+	// ErrServerNotSpecified returned when Server is not specified.
+	ErrServerNotSpecified = errors.New("server not present")
+)
+
+type ProcessorSink interface {
+	WriteSeries(database, retentionPolicy, name string, tags map[string]string, timestamp time.Time, values map[string]interface{}) error
+	DefaultRetentionPolicy(database string) (*influxdb.RetentionPolicy, error)
+}
+
+// Processor performs processing of Graphite data, and writes it to a sink.
+type Processor struct {
+	// Sink is the destination for processed Graphite data.
+	sink dataSink
+
+	// Database is the name of the database to insert data into.
+	Database string
+	// NamePosition is the position of name to be parsed from metric_path
+	NamePosition string
+	// NameSeparator is separator to parse metric_path with to get name and series
+	NameSeparator string
+}
+
+func NewProcessor(sink *ProcessorSink) *Processor {
+	p = Processor{}
+	return &p
+}
+
+// handleMessage decodes a graphite message from the reader and sends it to the
+// committer goroutine.
+func (p *processor) handleMessage(r *bufio.Reader) error {
+	// Decode graphic metric.
+	m, err := DecodeMetric(r, s.NamePosition, s.NameSeparator)
+	if err != nil {
+		return err
+	}
+
+	// Convert metric to a field value.
+	var values = make(map[string]interface{})
+	values[m.Name] = m.Value
+
+	retentionPolicy, err := s.sink.DefaultRetentionPolicy(s.Database)
+
+	if err != nil {
+		return fmt.Errorf("error looking up default database retention policy: %s", err)
+	}
+
+	if err := s.sink.WriteSeries(
+		s.Database,
+		retentionPolicy.Name,
+		m.Name,
+		m.Tags,
+		m.Timestamp,
+		values,
+	); err != nil {
+		return fmt.Errorf("write series data: %s", err)
+	}
+	return nil
+}
```

Comment:
I don't think you need a receiver on this function. If you store the `SeriesWriter` on the `TCPServer` and `UDPServer` then they can simply pass it in:

```
func handleMessage(w SeriesWriter, r *bufio.Reader, database, position, separator string) error {
	// Decode graphic metric.
	m, err := DecodeMetric(r, position, separator)
	if err != nil {
		return err
	}

	// Write series data.
	if err := w.WriteSeries(
		database,
		"",
		m.Name,
		m.Tags,
		m.Timestamp,
		map[string]interface{}{m.Name: m.Value},
	); err != nil {
		return fmt.Errorf("write series data: %s", err)
	}
	return nil
}
```